File: backend/app/clients/federal_register.py
import logging
from typing import Optional
from datetime import datetime, timedelta

from .base import BaseAPIClient
from ..config import get_settings
from ..models.schemas import SourceItem

logger = logging.getLogger(__name__)


class FederalRegisterClient(BaseAPIClient):
    def __init__(self):
        settings = get_settings()
        super().__init__(base_url=settings.federal_register_base_url)

    def _normalize_document(self, doc: dict) -> SourceItem:
        return SourceItem(
            source_type="federal_register",
            id=doc.get("document_number", ""),
            title=doc.get("title", "Untitled Document"),
            agency=", ".join(
                a.get("name", "")
                for a in doc.get("agencies", [])
                if a.get("name")
            ) or None,
            date=doc.get("publication_date"),
            url=doc.get("html_url", ""),
            excerpt=doc.get("abstract"),
            pdf_url=doc.get("pdf_url"),
            content_type=doc.get("type"),
            raw=doc,
        )

    async def search_documents(
        self,
        query: str,
        document_type: Optional[str] = None,
        agency: Optional[str] = None,
        days: Optional[int] = None,
        per_page: int = 10,
        page: int = 1,
    ) -> tuple[list[dict], list[SourceItem]]:
        params = {
            "conditions[term]": query,
            "per_page": min(per_page, 1000),
            "page": page,
            "order": "newest",
        }

        if document_type:
            params["conditions[type][]"] = document_type

        if agency:
            params["conditions[agencies][]"] = agency

        if days:
            end_date = datetime.now()
            start_date = end_date - timedelta(days=days)
            params["conditions[publication_date][gte]"] = start_date.strftime("%Y-%m-%d")

        url = f"{self.base_url}/documents.json"
        logger.info(f"Searching Federal Register: {query}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            params=params,
        )

        documents = data.get("results", [])
        sources = [self._normalize_document(doc) for doc in documents]

        return documents, sources

    async def get_document(self, document_number: str) -> tuple[Optional[dict], Optional[SourceItem]]:
        url = f"{self.base_url}/documents/{document_number}.json"
        logger.info(f"Fetching Federal Register document: {document_number}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
        )

        if data:
            return data, self._normalize_document(data)
        return None, None

File: backend-ts/src/clients/federalRegister.ts
import { BaseAPIClient } from "./base.js";
import { getSettings } from "../config.js";
import type { SourceItem } from "../models/schemas.js";

export class FederalRegisterClient extends BaseAPIClient {
    constructor() {
        const settings = getSettings();
        super(settings.federalRegisterBaseUrl);
    }

    private normalizeDocument(doc: Record<string, unknown>): SourceItem {
        const agencies = (doc.agencies || []) as { name?: string }[];
        const agencyNames = agencies
            .map((a) => a.name)
            .filter(Boolean)
            .join(", ");

        return {
            source_type: "federal_register",
            id: (doc.document_number as string) || "",
            title: (doc.title as string) || "Untitled Document",
            agency: agencyNames || null,
            date: doc.publication_date as string || null,
            url: doc.html_url as string || "",
            excerpt: doc.abstract as string || null,
            pdf_url: doc.pdf_url as string || null,
            content_type: doc.type as string || null,
            raw: doc,
        };
    }

    async searchDocuments(options: {
        query: string;
        documentType?: string;
        agency?: string;
        days?: number;
        perPage?: number;
        page?: number;
    }): Promise<{ documents: Record<string, unknown>[]; sources: SourceItem[] }> {
        const {
            query,
            documentType,
            agency,
            days,
            perPage = 10,
            page = 1,
        } = options;

        const params: Record<string, string | number> = {
            "conditions[term]": query,
            per_page: Math.min(perPage, 1000),
            page,
            order: "newest",
        };

        if (documentType) {
            params["conditions[type][]"] = documentType;
        }

        if (agency) {
            params["conditions[agencies][]"] = agency;
        }

        if (days) {
            const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000)
                .toISOString()
                .split("T")[0];
            params["conditions[publication_date][gte]"] = startDate;
        }

        const url = `${this.baseUrl}/documents.json`;

        const data = await this.requestWithRetry<{ results: Record<string, unknown>[] }>({
            url,
            params,
        });

        const documents = data.results || [];
        const sources = documents.map((doc) => this.normalizeDocument(doc));

        return { documents, sources };
    }

    async getDocument(
        documentNumber: string
    ): Promise<{ document: Record<string, unknown> | null; source: SourceItem | null }> {
        const url = `${this.baseUrl}/documents/${documentNumber}.json`;

        const data = await this.requestWithRetry<Record<string, unknown>>({
            url,
        });

        if (data) {
            return { document: data, source: this.normalizeDocument(data) };
        }
        return { document: null, source: null };
    }
}
